### ISSUE TITLE
Interrupt enable/disable should be using `compiler_fence`.

### ISSUE THREAD

**Issue by cr1901** [rust-embedded/msp430#11]
For example, in [`cortex_m`](https://github.com/rust-embedded/cortex-m/blob/master/src/interrupt.rs#L39):

```
pub fn disable() {
    unsafe {
        asm!("cpsid i", options(nomem, nostack, preserves_flags));
    }

    // Ensure no subsequent memory accesses are reordered to before interrupts are disabled.
    compiler_fence(Ordering::SeqCst);
}
```

`compiler_fence` [does not currently work properly](https://github.com/rust-lang/rust/issues/27731#issuecomment-612362819) for the msp430 backend. Thank you @eddyb for pointing this out to me.

[`barrier` and `nop`](https://github.com/rust-embedded/msp430/blob/master/src/asm.rs) are also likely affected; `barrier` needs a fence to prevent reordering, and both functions assume that Rust will not optimize out inline ASM (which I'm not sure is true).

**Comment by taiki-e** [comment]
IIUC, inline assembly with neither `nomem` nor `readonly` implies a (SeqCst) compiler fence ([refs](https://rust-lang.zulipchat.com/#narrow/stream/216763-project-inline-asm/topic/asm.20and.20compiler.20fences/near/274118488)), so, the current usage in this crate should not be affected.
https://github.com/rust-embedded/msp430/blob/4ee7a8b461763feac33409ff67fb6df1a41142a1/src/interrupt.rs#L13
https://github.com/rust-embedded/msp430/blob/4ee7a8b461763feac33409ff67fb6df1a41142a1/src/interrupt.rs#L32
https://github.com/rust-embedded/msp430/blob/4ee7a8b461763feac33409ff67fb6df1a41142a1/src/asm.rs#L9
https://github.com/rust-embedded/msp430/blob/4ee7a8b461763feac33409ff67fb6df1a41142a1/src/asm.rs#L17

**Comment by cr1901** [comment]
@taiki-e Okay, good to know (thank you for the link as well). Also, from what I understand since I wrote the issue, inline `asm!` won't ever be optimized _out_ either unless it's given the `pure` flag. Since none of the `asm!` snippets in this crate use flags, we should be good.

However, `compiler_fence` really does need to be implemented properly for msp430 (at which point I can make `barrier` a call to `compiler_fence(Ordering::SeqCst)` for backwards compat). I'll probably change the title in the morning.

**Comment by taiki-e** [comment]
> inline `asm!` won't ever be optimized _out_ either unless it's given the `pure` flag

~~IIUC, this is due to ["LLVM doesn't have attributes to represent ReadOnly/ReadNone + SideEffect"](https://github.com/rust-lang/rust/blob/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/compiler/rustc_codegen_llvm/src/asm.rs#L279-L290).~~ EDIT: see https://github.com/rust-embedded/msp430/issues/11#issuecomment-1204928442
~~Non-LLVM backends or LLVM in the future may support this, and if you use them, code that relies on the current LLVM behavior may run into problems.~~

EDIT2: I seem to have misunderstood your comment: https://github.com/rust-embedded/msp430/issues/11#issuecomment-1205534416

**Comment by taiki-e** [comment]
> IIUC, this is due to "LLVM doesn't have attributes to represent ReadOnly/ReadNone + SideEffect"

Oh, considering that LLVM has [InaccessibleMemOnly (ReadNone + SideEffect)](https://github.com/rust-lang/rust/blob/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/compiler/rustc_codegen_llvm/src/asm.rs#L286-L287), my understanding here is wrong.

(Is it actually because it is [volatile](https://github.com/rust-lang/rust/blob/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/compiler/rustc_codegen_llvm/src/asm.rs#L248)?)

**Comment by eddyb** [comment]
To be clear I don't/didn't actually know what LLVM guarantees for inline assembly, I couldn't find any ordering interactions being described but I wouldn't be surprised to hear that (without `nomem`) it's equivalent to `SeqCst`.

**Comment by cr1901** [comment]
>Is it actually because it is volatile?

Yes, volatile ("Do not optimize away. Do not reorder with other volatile code.") was what I had in mind.

I don't see why `ReadOnly/ReadNone + SideEffect` would allow removing `asm` block offhand/can't think of example. But noted that maybe `pure` alone determining volatility is subject to change.

**Comment by taiki-e** [comment]
> I don't see why `ReadOnly/ReadNone + SideEffect` would allow removing `asm` block offhand/can't think of example. But noted that maybe `pure` alone determining volatility is subject to change.

volatile seems to guarantee that `asm` will not be removed, but volatile itself does not guarantee that (EDIT: non-volatile) memory accesses will not be reordered (compiler_fence guarantees it. EDIT: asm without `nomem` and `readonly` also guarantees it).

`ReadNone + SideEffect` allows reordering (and omitting) of preceding and subsequent memory accesses because it ensures that `asm` does not perform memory accesses: [godbolt](https://godbolt.org/z/Yhzva4cve)

As for `ReadOnly + SideEffect`, I think it should allow reordering of preceding and subsequent memory reads because it ensures that `asm` does not modify memory. (However, [LLVM does not currently seem to support this optimization](https://github.com/rust-lang/rust/blob/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/compiler/rustc_codegen_llvm/src/asm.rs#L289))

**Comment by cr1901** [comment]
>volatile seems to guarantee that asm will not be removed, but it does not guarantee that memory accesses will not be reordered

I was confused because in a [previous comment](https://github.com/rust-embedded/msp430/issues/11#issuecomment-1204919256) you said:

>Non-LLVM backends or LLVM in the future may support this, and if you use them, code that relies on the current LLVM behavior may run into problems.

_Even with your correction in the next comment_, I interpreted your comment to mean:

"Right now, LLVM will treat inline asm as volatile as long as you don't use `pure`. However in future LLVM and other backends, the rules for whether inline asm is treated as volatile will be more intricate than 'was `pure` specified or not?'". An example of `asm` that LLVM may _not_ treat as volatile in the future is ReadOnly/ReadNone + SideEffect. _I can't think of an example offhand where LLVM could reasonably optimize out asm code with side effects (`InaccessibleMemOnly`- how would LLVM be able to examine an `asm` code snippet to find out the set of memory locations actually modified?), but that doesn't mean they don't exist._

The above shouldn't affect this crate.

>compiler_fence guarantees it

Additionally, as [established](https://github.com/rust-embedded/msp430/issues/11#issuecomment-1204704916) writing `asm` without `nomem` and `readonly` implies a fence. At least I hope, because otherwise my `asm` as written in this crate is wrong :P.

**Comment by taiki-e** [comment]
> _Even with your correction in the next comment_, I interpreted your comment to mean:
>
> "Right now, LLVM will treat inline asm as volatile as long as you don't use pure. However in future LLVM and other backends, the rules for whether inline asm is treated as volatile will be more intricate than 'was pure specified or not?'".

Sorry, I thought your comment ("inline asm! won't ever be optimized ..." in https://github.com/rust-embedded/msp430/issues/11#issuecomment-1204723519) was not discussing "can asm be removed", but rather regarding "does it allow reordering preceding and subsequent memory accesses" (which is a requirement for using it as a compiler fence). If I read it as a discussion of the former, my comments are indeed very confusing.
